src/bun.js/http.exports.js

```diff
@@ -0,0 +1,480 @@
+    const headers = Object.create(null);
+    const no_body = 'GET' === method || 'HEAD' === method;
+
+    for (const key of req.headers.keys()) {
+      const value = req.headers.get(key);
+
+      headers[key] = value;
+      rawHeaders.push(key, value);
+    }
+
+    const read_body = async (size) => {
+      // TODO: streaming
+      try {
+        body = Buffer.from(await req.arrayBuffer());
+        this.push(body.subarray(body_offset, body_offset += size));
+      } catch (err) {
+        this.destroy(err);
+      }
+    };
+
+    super({
+      read: size => {
+        if (no_body) {
+          this.push(null);
+          this.complete = true;
+        }
+
+        else {
+          if (!body) read_body(size);
+
+          else {
+            if (body_offset >= body.length) {
+              this.push(null);
+              this.complete = true;
+            } else {
+              this.push(body.subarray(body_offset, body_offset += size));
+            }
+          }
+        }
+      },
+    });
+
+    const url = new URL(req.url);
+    // TODO: reuse trailer object?
+    // TODO: get hostname and port from Bun.serve and calculate substring() offset
+
+    this.complete = false;
+    this.rawTrailers = [];
+    this.headers = headers;
+    this.httpVersion = '1.1';
+    this.method = req.method;
+    this.httpVersionMajor = 1;
+    this.httpVersionMinor = 1;
+    this.rawHeaders = rawHeaders;
+    this.trailers = Object.create(null);
+    this.url = url.pathname + url.search;
+  }
+
+  get aborted() {
+    throw new Error('not implemented');
+  }
+
+  get connection() {
+    throw new Error('not implemented');
+  }
+
+  get statusCode() {
+    throw new Error('not implemented');
+  }
+
+  get statusMessage() {
+    throw new Error('not implemented');
+  }
+
+  get socket() {
+    const duplex = new EventEmitter();
+    this.on('end', () => duplex.emit('end'));
+    this.on('close', () => duplex.emit('close'));
+
+    return duplex;
+  }
+
+  setTimeout(msecs, callback) {
+    throw new Error('not implemented');
+  }
+}
+
+export class ServerResponse extends Writable {
+  constructor({ req, reply }) {
+    const headers = new Headers;
+    const sink = new Bun.ArrayBufferSink();
+
+    super({
+      construct: callback => {
+        sink.start({
+          stream: false,
+          asUint8Array: true,
+        });
+
+        callback();
+      },
+
+      write: (chunk, encoding, callback) => {
+        this.headersSent = true;
+
+        sink.write(chunk);
+
+        callback();
+      },
+
+      writev: (chunks, callback) => {
+        this.headersSent = true;
+
+        for (const chunk of chunks) {
+          sink.write(chunk.chunk);
+        }
+
+        callback();
+      },
+
+      final: callback => {
+        callback();
+        this.headersSent = true;
+
+        if (this.sendDate && !headers.has('date')) {
+          headers.set('date', new Date().toUTCString());
+        }
+
+        reply(new Response(sink.end(), {
```

Comment:
This does another allocation to clone the Uint8Array, IIRC

wonder if we should have a version of this that doesn't do that

could add an extremely messy flag to ArrayBufferSink which...returns a Blob